include/deal.II/numerics/fe_field_function.templates.h

```diff
@@ -438,162 +442,19 @@ namespace Functions
-    unsigned int c = 0;
-
-    while (left_over == true)
-      {
-        // Assume this is the last one
-        left_over = false;
-        Assert(first_outside < np,
-               ExcIndexRange(first_outside, 0, np));
-
-        // If we found one in this cell, keep looking in the same cell
-        for (unsigned int p=first_outside; p<np; ++p)
-          if (point_flags[p] == false)
-            {
-              // same logic as above
-              const boost::optional<Point<dim> >
-              qp = get_reference_coordinates (cells[c], points[p]);
-              if (qp)
-                {
-                  point_flags[p] = true;
-                  qpoints[c].push_back(qp.get());
-                  maps[c].push_back(p);
-                }
-              else
-                {
-                  // Set things up for next round
-                  if (left_over == false)
-                    first_outside = p;
-                  left_over = true;
-                }
-            }
-        // If we got here and there is
-        // no left over, we are
-        // done. Else we need to find
-        // the next cell
-        if (left_over == true)
-          {
-            const std::pair<typename dealii::internal::ActiveCellIterator<dim, dim, DoFHandlerType>::type, Point<dim> > my_pair
-              = GridTools::find_active_cell_around_point (mapping, *dh, points[first_outside]);
-            AssertThrow (!my_pair.first->is_artificial(),
-                         VectorTools::ExcPointNotAvailableHere());
-
-            cells.push_back(my_pair.first);
-            qpoints.emplace_back(1, my_pair.second);
-            maps.emplace_back(1, first_outside);
-            c++;
-            point_flags[first_outside] = true;
-            // And check if we can exit the loop now
-            if (first_outside == np-1)
-              left_over = false;
-          }
-      }
-
-    // Augment of one the number of cells
-    ++c;
-    // Debug Checking
-    Assert(c == cells.size(), ExcInternalError());
-
-    Assert(c == maps.size(),
-           ExcDimensionMismatch(c, maps.size()));
-
-    Assert(c == qpoints.size(),
-           ExcDimensionMismatch(c, qpoints.size()));
-
-#ifdef DEBUG
-    unsigned int qps = 0;
-    // The number of points in all
-    // the cells must be the same as
-    // the number of points we
-    // started off from.
-    for (unsigned int n=0; n<c; ++n)
-      {
-        Assert(qpoints[n].size() == maps[n].size(),
-               ExcDimensionMismatch(qpoints[n].size(), maps[n].size()));
-        qps += qpoints[n].size();
-      }
-    Assert(qps == np,
-           ExcDimensionMismatch(qps, np));
-#endif
-
-    return c;
+    // Getting the reference to the underlying triangulation and initializing the cache
+    const Triangulation<dim,DoFHandlerType::space_dimension> &tria = dh->get_triangulation();
+    GridTools::Cache<dim,DoFHandlerType::space_dimension> cache(tria,mapping);
```

Comment:
Can you add a member `cache` to the `FEFieldFunction` class, initialized at construction time? In this way you don't recompute the cache at each round.